src/ser.rs

```diff
@@ -0,0 +1,1130 @@
+    type Ok = T;
+    type Error = Error;
+
+    fn serialize_field<U: ?Sized>(&mut self, _value: &U) -> Result<(), Error>
+        where U: ser::Serialize,
+    {
+        match self.nope {}
+    }
+
+    fn end(self) -> Result<T, Error> {
+        match self.nope {}
+    }
+}
+
+impl<T> ser::SerializeTupleVariant for Impossible<T> {
+    type Ok = T;
+    type Error = Error;
+
+    fn serialize_field<U: ?Sized>(&mut self, _value: &U) -> Result<(), Error>
+        where U: ser::Serialize,
+    {
+        match self.nope {}
+    }
+
+    fn end(self) -> Result<T, Error> {
+        match self.nope {}
+    }
+}
+
+impl<T> ser::SerializeStructVariant for Impossible<T> {
+    type Ok = T;
+    type Error = Error;
+
+    fn serialize_field<U: ?Sized>(&mut self, _key: &'static str, _value: &U)
+                                  -> Result<(), Error>
+        where U: ser::Serialize,
+    {
+        match self.nope {}
+    }
+
+    fn end(self) -> Result<T, Error> {
+        match self.nope {}
+    }
+}
+
+struct StringExtractor<'a> {
+    slot: &'a mut String,
+}
+
+impl<'a> ser::Serializer for StringExtractor<'a> {
+    type Ok = ();
+    type Error = Error;
+    type SerializeSeq = Impossible<()>;
+    type SerializeTuple = Impossible<()>;
+    type SerializeTupleStruct = Impossible<()>;
+    type SerializeTupleVariant = Impossible<()>;
+    type SerializeMap = Impossible<()>;
+    type SerializeStruct = Impossible<()>;
+    type SerializeStructVariant = Impossible<()>;
+
+    fn serialize_bool(self, _v: bool) -> Result<(), Self::Error> {
+        Err(Error::KeyNotString)
+	}
+
+    fn serialize_i8(self, _v: i8) -> Result<(), Self::Error> {
+        Err(Error::KeyNotString)
+    }
+
+    fn serialize_i16(self, _v: i16) -> Result<(), Self::Error> {
+        Err(Error::KeyNotString)
+    }
+
+    fn serialize_i32(self, _v: i32) -> Result<(), Self::Error> {
+        Err(Error::KeyNotString)
+    }
+
+    fn serialize_i64(self, _v: i64) -> Result<(), Self::Error> {
+        Err(Error::KeyNotString)
+    }
+
+    fn serialize_u8(self, _v: u8) -> Result<(), Self::Error> {
+        Err(Error::KeyNotString)
+    }
+
+    fn serialize_u16(self, _v: u16) -> Result<(), Self::Error> {
+        Err(Error::KeyNotString)
+    }
+
+    fn serialize_u32(self, _v: u32) -> Result<(), Self::Error> {
+        Err(Error::KeyNotString)
+    }
+
+    fn serialize_u64(self, _v: u64) -> Result<(), Self::Error> {
+        Err(Error::KeyNotString)
+    }
+
+    fn serialize_f32(self, _v: f32) -> Result<(), Self::Error> {
+        Err(Error::KeyNotString)
+    }
+
+    fn serialize_f64(self, _v: f64) -> Result<(), Self::Error> {
+        Err(Error::KeyNotString)
+    }
+
+    fn serialize_char(self, _v: char) -> Result<(), Self::Error> {
+        Err(Error::KeyNotString)
+    }
+
+    fn serialize_str(self, value: &str) -> Result<(), Self::Error> {
+        if value.contains("\n") {
+            return Err(Error::KeyNewline)
+        }
+        assert_eq!(self.slot.len(), 0);
```

Comment:
I would recommend `assert!(self.slot.is_empty())` but with the suggestion above we don't need this assertion anymore. `Ok(value.to_owned())`.